Change _WKWebExtension creation methods to be asynchronous.

xeenon opened the pull request:
#### 8b43053675f985642ddcad0b88f0bcf0219ad7c8
<pre>
Change _WKWebExtension creation methods to be asynchronous.
<a href="https://webkit.org/b/276196">https://webkit.org/b/276196</a>
<a href="https://rdar.apple.com/problem/131076684">rdar://problem/131076684</a>

Reviewed by Jeff Miller and Alex Christensen.

Add asynchronous convenience methods to _WKWebExtension so we can update these
to load the extension data on a background thread in the future (see bug 276194).

Moved the existing init methods to _WKWebExtensionPrivate.h.

Also make sure the watchOS and tvOS stubs always return an unsupported error and
call the completion handlers.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
(+[_WKWebExtension extensionWithAppExtensionBundle:completionHandler:]): Added.
(+[_WKWebExtension extensionWithResourceBaseURL:completionHandler:]): Added.
(-[_WKWebExtension initWithAppExtensionBundle:error:]): Call the private method.
(-[_WKWebExtension _initWithAppExtensionBundle:error:]): Added.
(-[_WKWebExtension initWithResourceBaseURL:error:]): Call the private method.
(-[_WKWebExtension _initWithResourceBaseURL:error:]): Added.
(+[_WKWebExtension extensionWithAppExtensionBundle:]): Deleted.
(+[_WKWebExtension extensionWithResourceBaseURL:]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext loadBackgroundContentWithCompletionHandler:]): Call completionHandler
with unsupported error.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.mm:
(-[_WKWebExtensionController loadExtensionContext:error:]): Return unsupported error.
(-[_WKWebExtensionController unloadExtensionContext:error:]): Ditto.
(-[_WKWebExtensionController fetchDataRecordsOfTypes:completionHandler:]): Call completionHandler.
(-[_WKWebExtensionController fetchDataRecordOfTypes:forExtensionContext:completionHandler:]): Ditto.
(-[_WKWebExtensionController removeDataOfTypes:forDataRecords:completionHandler:]): Ditto.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionMatchPattern.mm:
(-[_WKWebExtensionMatchPattern initWithString:error:]): Return unsupported error.
(-[_WKWebExtensionMatchPattern initWithScheme:host:path:error:]): Ditto.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionMessagePort.mm:
(-[_WKWebExtensionMessagePort sendMessage:completionHandler:]): Call completionHandler.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionPrivate.h: Moved init method here.
Added missing documentation comments.

Canonical link: <a href="https://commits.webkit.org/280642@main">https://commits.webkit.org/280642@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/39ad1041454ee1e4273fbaccba293dfffdccce96

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57191 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36519 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9666 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60812 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7634 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44143 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7824 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/46309 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5373 "Passed tests") 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59221 "Failed to checkout and rebase branch from PR 30461") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/34273 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49379 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27169 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/31055 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6639 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53010 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/6965 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62492 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1104 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/7063 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/53568 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1109 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49418 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/53636 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/930 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8531 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/32348 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/33433 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/34518 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33179 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
```diff
@@ -44,35 +44,55 @@ @implementation _WKWebExtension
-        return nil;
+    dispatch_async(dispatch_get_main_queue(), ^{
+        NSError * __autoreleasing error;
+        Ref result = WebKit::WebExtension::create(appExtensionBundle, &error);
```
I think it would be a little nicer to make WebExtension::create return an Expected&lt;Ref&lt;WebExtension&gt;, RetainPtr&lt;NSError&gt;&gt; but that's definitely unrelated to this PR.

xeenon replied:
Agreed! I'll note that for another change.

jeffmapple commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
```diff
@@ -298,22 +324,30 @@ - (BOOL)_backgroundContentUsesModules
-+ (instancetype)extensionWithAppExtensionBundle:(NSBundle *)appExtensionBundle
++ (void)extensionWithAppExtensionBundle:(NSBundle *)appExtensionBundle completionHandler:(void (^)(_WKWebExtension *extension, NSError *error))completionHandler
 {
-    return nil;
```
I understand that these methods are no-ops when web extension support isn't enabled, but shouldn't we still call the completion handler here and in `+ extensionWithResourceBaseURL`?

xeenon replied:
Yeah, I can do that. I was on the fence about it too.

xeenon committed:
Change _WKWebExtension creation methods to be asynchronous.
https://webkit.org/b/276196
rdar://problem/131076684

Reviewed by Jeff Miller and Alex Christensen.

Add asynchronous convenience methods to _WKWebExtension so we can update these
to load the extension data on a background thread in the future (see bug 276194).

Moved the existing init methods to _WKWebExtensionPrivate.h.

Also make sure the watchOS and tvOS stubs always return an unsupported error and
call the completion handlers.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
(+[_WKWebExtension extensionWithAppExtensionBundle:completionHandler:]): Added.
(+[_WKWebExtension extensionWithResourceBaseURL:completionHandler:]): Added.
(-[_WKWebExtension initWithAppExtensionBundle:error:]): Call the private method.
(-[_WKWebExtension _initWithAppExtensionBundle:error:]): Added.
(-[_WKWebExtension initWithResourceBaseURL:error:]): Call the private method.
(-[_WKWebExtension _initWithResourceBaseURL:error:]): Added.
(+[_WKWebExtension extensionWithAppExtensionBundle:]): Deleted.
(+[_WKWebExtension extensionWithResourceBaseURL:]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext loadBackgroundContentWithCompletionHandler:]): Call completionHandler
with unsupported error.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.mm:
(-[_WKWebExtensionController loadExtensionContext:error:]): Return unsupported error.
(-[_WKWebExtensionController unloadExtensionContext:error:]): Ditto.
(-[_WKWebExtensionController fetchDataRecordsOfTypes:completionHandler:]): Call completionHandler.
(-[_WKWebExtensionController fetchDataRecordOfTypes:forExtensionContext:completionHandler:]): Ditto.
(-[_WKWebExtensionController removeDataOfTypes:forDataRecords:completionHandler:]): Ditto.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionMatchPattern.mm:
(-[_WKWebExtensionMatchPattern initWithString:error:]): Return unsupported error.
(-[_WKWebExtensionMatchPattern initWithScheme:host:path:error:]): Ditto.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionMessagePort.mm:
(-[_WKWebExtensionMessagePort sendMessage:completionHandler:]): Call completionHandler.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionPrivate.h: Moved init method here.
Added missing documentation comments.

Canonical link: https://commits.webkit.org/280642@main
